.circleci/build_image.sh

```diff
@@ -0,0 +1,31 @@
+#!/bin/sh
+set -xe
+
+if [[ $# -ne 2 ]]; then
+  echo "Usage: $0 DOCKERFILE IMAGE_NAME"
+  exit 2
+fi
+
+DIR=`dirname $1`
+DOCKERFILE=$1
+IMAGE_NAME=$2
+
+if [[ -z $CIRCLE_PULL_REQUEST ]]; then
+  # Log in to Docker Hub.
+  docker login -u $DOCKER_USER -p $DOCKER_PASS
+  # Decide whether Docker Hub images should be tagged ":branch-Y" or ":latest".
+  if [[ $CIRCLE_BRANCH != "master" ]]; then
+    DOCKERHUB_TAG="branch-$CIRCLE_BRANCH"
+  else
+    DOCKERHUB_TAG="latest"
+  fi
+fi
+
+# Pull published image for cached layers. (Allowed to fail)
+docker pull $IMAGE_NAME || true
```

Comment:
Nit: I don't think we should `docker pull` here:

- In the previous Circle CI configuration, we only pulled `janx/ubuntu-dev` to pass the image from one job to the next, and you're replacing this with `docker save`/`load`.
- We purposefully *didn't* cache project images, in order to ensure that builds continue to work from scratch and that repositories/sources are always the freshest possible.
- (We also pulled `ubuntu-dev` specifically in the `ubuntu-dev` job, in order to accelerate base image builds, but I'm ok with removing that call as well, in order to get fresher packages as well.)